### DISCUSSION TITLE
Is it possible to configure Reqnroll to run using STA (Single-Threaded Apartment)?

### DISCUSSION THREAD

**Discussion by Pressacco** [reqnroll/Reqnroll#279] [Q&A]
Dear Reqnroll community.

I am currently exploring the idea of combining _Reqnroll_ with `ViewModel` level testing of a C# WPF application.

# The Problem

## Approach 1 

Apparently you can configure MS Test to run in STA mode by simply:

```CSharp
using System.Threading;
using Microsoft.VisualStudio.TestTools.UnitTesting;

[TestClass]
public class TestInitialization
{
    [AssemblyInitialize]
    public static void AssemblyInit(TestContext context)
    {
        // Set the main thread to Single-Threaded Apartment
        Thread.CurrentThread.SetApartmentState(ApartmentState.STA);
    }
}
```

Problem:
- Unfortunately this does not work with _Reqnroll_, because `AssemblyInitialize` can only be referenced once & _Reqnroll_ uses this attribute in the generated _Reqnroll_ code behind.

## Approach 2

The following code compiles & runs as expected:

```CSharp
        // https://www.meziantou.net/mstest-v2-customize-test-execution.htm
        public class StaTestClassAttribute : TestClassAttribute
	{
		public override TestMethodAttribute GetTestMethodAttribute(TestMethodAttribute testMethodAttribute)
		{
			if (testMethodAttribute is StaTestMethodAttribute)
				return testMethodAttribute;

			return new StaTestMethodAttribute(base.GetTestMethodAttribute(testMethodAttribute));
		}
	}

        [StaTestClass]
	public class UnitTestClass
	{
		[TestMethod]
		public void UnitTestExample()
		{
                        // Use NSubstitute to create a mock window
			var mainWindow = Substitute.For<Window>();

                        // pass in the Window dependency
                        var  viewModel = new DashboardViewModel(mainWindow);

                       // act
                       
                       // assert
                }
       }
```

Problem:
- When this line `Substitute.For<Window>()` is called from a _Reqnroll_ step, .NET Core 8 throws an exception:
   - `InvalidOperationException` The calling thread must be STA, because many UI components require this.

# Question

Is there an easy way to configure _Reqnroll_ to run in STA mode?  Any guidance you can provide would be appreciated.

# Context

- .NET Core 6
   - WPF application
- MS Test 3.6.0
- Reqnroll 2.1.0

# Related Discussions

- #133

---

It would be nice if the view model only referenced other view models (i.e. had no concept of a WPF `Window`), but let's assume that I do not have control over this implementation detail.

**Comment by gasparnagy** [comment]
Maybe you could do it with an in-project Reqnroll plugin. You would only need to add this to your project (not tested):

```
﻿using Reqnroll.BoDi;
using Reqnroll.Infrastructure;
using Reqnroll.Plugins;
using Reqnroll.UnitTestProvider;

[assembly: RuntimePlugin(typeof(STAPlugin ))]

namespace YourNamespace;

public class STAPlugin : IRuntimePlugin
{
    public void Initialize(RuntimePluginEvents runtimePluginEvents, RuntimePluginParameters runtimePluginParameters, UnitTestProviderConfiguration unitTestProviderConfiguration)
    {
        // this code is called from the assembly initialize, so doing it simply here could work:
        Thread.CurrentThread.SetApartmentState(ApartmentState.STA);

        runtimePluginEvents.CustomizeGlobalDependencies += (sender, args) =>
        {
            // if the above does not work, you can also try it here
        };
    }
}
```

**Comment by Pressacco** [comment]
Unfortunately calling `Thread.CurrentThread.SetApartmentState(ApartmentState.STA);` anywhere within `IRuntimePlugin` results in an exception being thrown:
```C#
InvalidOperationException: Failed to set the specified COM apartment state. Current apartment state 'MTA'.
```

Ideally in MVVM, the `ViewModel` shouldn't know about a `Window`. I am going to try to refactor my code in order to remove the `Window` references.

Thank you @gasparnagy for taking the time to respond to my question.
